parser_api/example.rb

```diff
@@ -0,0 +1,173 @@
+require 'puppet'
+
+# Example of a module setting everything up to perform custom
+# validation of an AST model produced by parsing puppet source.
+#
+module MyValidation
+
+  # A module for the new issues that the this new kind of validation will generate
+  #
+  module Issues
+    # (see Puppet::Pops::Issues#issue)
+    # This is boiler plate code
+    def self.issue (issue_code, *args, &block)
+      Puppet::Pops::Issues.issue(issue_code, *args, &block)
+    end
+
+    INVALID_WORD = issue :INVALID_WORD, :text do
+      "The word '#{text}' is not a real word."
+    end
+  end
+
+  # This is the class that performs the actual validation by checking input
+  # and sending issues to an acceptor.
+  #
+  class MyChecker
+    attr_reader :acceptor
+    def initialize(diagnostics_producer)
+      @@bad_word_visitor       ||= Puppet::Pops::Visitor.new(nil, "badword", 0, 0)
+      # add more polymorphic checkers here
+
+      # remember the acceptor where the issues should be sent
+      @acceptor = diagnostics_producer
+    end
+
+    # Validates the entire model by visiting each model element and calling the various checkers
+    # (here just the example 'check_bad_word'), but a series of things could be checked.
+    #
+    # The result is collected by the configured diagnostic provider/acceptor
+    # given when creating this Checker.
+    # 
+    # Returns the @acceptor for convenient chaining of operations
+    #
+    def validate(model)
+      # tree iterate the model, and call the checks for each element
+
+      # While not strictly needed, here a check is made of the root (the "Program" AST object)
+      check_bad_word(model)
+
+      # Then check all of its content
+      model.eAllContents.each {|m| check_bad_word(m) }
```

Comment:
Don't think we have an alternative to eAllContents that is available from 3.8 future parser and forward?
And I think we should say that only eAllContent is API to reduce the amount of exposure to RGen API that we may not want to keep in an optimized implementation,